### ISSUE TITLE
Build using emcc failed

### ISSUE THREAD

**Issue by ChinYikMing** [sysprog21/rv32emu#375]
After this [commit](https://github.com/sysprog21/rv32emu/commit/f484a4b92cd7b26779f44613b973fe1be83511aa), `emcc` fails to build the source. On my machine, the command in [mk/toolchain.mk](https://github.com/sysprog21/rv32emu/blob/master/mk/toolchain.mk): `emcc --version | head -n 1` will output `emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 3.1.51 (c0c2ca1314672a25699846b4663701bcb6f69cca)`. The output contains the keywords `clang` and `emcc`, thus both `CC_IS_CLANG` and `CC_IS_EMCC` variables will be set. Consequently, the `CFLAGS` from `clang` and `emcc` will be mixed. This causes the build to fail.

Note that supposed the build machine has installed both `emcc` and `clang`

**Comment by ChinYikMing** [comment]
May I know the concern of the way to check which compiler was used in [mk/toolchain.mk](https://github.com/sysprog21/rv32emu/blob/master/mk/toolchain.mk) of [commit](https://github.com/sysprog21/rv32emu/commit/959cd1769384989236d2a3189788191f4991757b) ? @jserv

**Comment by jserv** [comment]
The commit f484a4b92cd7b26779f44613b973fe1be83511aa was used as the workaround for Apple LLVM toolchain, and the above resulted from incorrect detection of toolchain built upon clang or gcc.

**Comment by ChinYikMing** [comment]
> The commit [f484a4b](https://github.com/sysprog21/rv32emu/commit/f484a4b92cd7b26779f44613b973fe1be83511aa) was used as the workaround for Apple LLVM toolchain, and the above resulted from incorrect detection of toolchain built upon clang or gcc.

The version of Apple LLVM toolchain of my machine is `Homebrew clang version 17.0.6` 
(run `clang --version | head -n 1`) and I am able to detect `clang` and build rv32emu with it. May I know the Apple LLVM toolchain of your machine?

**Comment by jserv** [comment]
> The version of Apple LLVM toolchain of my machine is `Homebrew clang version 17.0.6` (run `clang --version | head -n 1`) and I am able to detect `clang` and build rv32emu with it. May I know the Apple LLVM toolchain of your machine?

The workaround was applied to  Xcode 15.
See commit 618d4f156d8bc69d38898d72aab7aed51d43b154

**Comment by ChinYikMing** [comment]
I think we are referring to different things. Let me rephrase the question:

May I know the concern of changing from this
```
ifneq ($(shell $(CC) --version | head -n 1 | grep emcc),)
     CC_IS_EMCC := 1
else ifneq ($(shell $(CC) --version | head -n 1 | grep clang),)
     CC_IS_CLANG := 1
else
     ifneq ($(shell $(CC) --version | grep "Free Software Foundation"),)
         CC_IS_GCC := 1
     endif
endif
```

to this in this commit [f484a4b](https://github.com/sysprog21/rv32emu/commit/f484a4b92cd7b26779f44613b973fe1be83511aa)?
```
ifneq ($(shell $(CC) --version | head -n 1 | grep emcc),)
     CC_IS_EMCC := 1
endif
ifneq ($(shell $(CC) --version | head -n 1 | grep clang),)
     CC_IS_CLANG := 1
endif
ifneq ($(shell $(CC) --version | grep "Free Software Foundation"),)
     CC_IS_GCC := 1
endif
```

Note that the former will successfully build with `emcc`.

**Comment by jserv** [comment]
I aimed to simplify the process by treating `emcc` as a cross-compiler, thereby tweaking the build rules. If you have any proposed changes, please don't hesitate to submit them.
